sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ExecutionStateSampler.java

```diff
@@ -169,7 +198,16 @@ private Void stateSampler() throws Exception {
         long millisSinceLastSample = currentTimeMillis - lastSampleTimeMillis;
         synchronized (activeStateTrackers) {
           for (ExecutionStateTracker activeTracker : activeStateTrackers) {
-            activeTracker.takeSample(currentTimeMillis, millisSinceLastSample);
+            Optional<String> errMsg =
+                activeTracker.takeSample(currentTimeMillis, millisSinceLastSample);
+            if (errMsg.isPresent() && this.onTimeoutExceededCallback != null) {
+              this.onTimeoutExceededCallback.accept(
```

Comment:
Probably safer to move this out of the synchronized block since we're calling out to other code and we don't want possible deadlocks.

Optional<String> timeoutMsg = Optional.empty();
synchronized (activeStateTrackers) {
          for (ExecutionStateTracker activeTracker : activeStateTrackers) {
            timeoutMsg = timeoutMsg.or(activeTracker.takeSample(currentTimeMillis, millisSinceLastSample);
          }
       }
if (timeoutMsg.isPresent() && this.onTimeoutExceededCallback != null) {
              this.onTimeoutExceededCallback.accept(...)
}